Improve citus lock waits to show non-tx blocked distributed processes as well

onderkalaci opened the pull request:
Fixes https://github.com/citusdata/citus/issues/5287

Each individual commit has its own explanation. Basically, rely on global_pid infrastructure and how any blocked distributed processes. Before these changes, we were only showing blocked processes that are inside a distributed transaction.

marcocitus commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -124,9 +133,14 @@ BuildGlobalWaitGraph(void)
 
-		int querySent = SendRemoteCommand(connection, command);
+		appendStringInfo(queryString,
+						 "SELECT * FROM dump_local_wait_edges(%s)",
```
It's a pity we don't use an explicit column list here. We'll break rolling upgrades by returning additional columns.

I kind of wonder whether we should rename the function and keep the old signature.

onderkalaci replied:
I ended up doing that

onderkalaci commented on src/test/regress/expected/isolation_rebalancer_deferred_drop.out:
```diff
@@ -148,7 +148,8 @@ step s1-move-placement-back:
     SELECT master_move_shard_placement((SELECT * FROM selected_shard), 'localhost', 57638, 'localhost', 57637);
-
+ <waiting ...>
+step s1-move-placement-back: <... completed>
```
I think so, `s2-lock-table-on-worker` should be holding a lock, so `master_move_shard_placement` would be blocked as far as I can follow the code

onderkalaci replied:
to clarify, bug is in the isolation test lock detection UDFs, not in the other UDFs

marcocitus commented on src/backend/distributed/sql/citus--10.2-4--11.0-1.sql:
```diff
@@ -17,7 +17,12 @@
 #include "udfs/get_all_active_transactions/11.0-1.sql"
 #include "udfs/get_global_active_transactions/11.0-1.sql"
 
+#include "udfs/dump_local_blocked_processes/11.0-1.sql"
```
does the dump prefix still make sense?

onderkalaci replied:
Probably not necessary, though given these are mostly used for `citus_lock_waits` and not very much intended for calling regularly, I didn't give much thought about this.

So, I may change it to ` local_blocked_processes` /   `global_blocked_processes`. Does that sound better?

onderkalaci replied:
Ok, given that these UDFs are low-level things that we only need for `citus_lock_waits` and isolation tests, I'm updating the names to: `citus_internal_local_blocked_processes` and `citus_internal_global_blocked_processes`.

We don't expect/want users to play with these low level UDFs as it contains the details that not many need (at all).

marcocitus commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -124,9 +173,20 @@ BuildGlobalWaitGraph(void)
+		if (onlyDistributedTx)
+		{
+			appendStringInfo(queryString,
+							 "SELECT * FROM dump_local_wait_edges()");
```
we should consider using explicit column names

marcocitus commented:
Something weird is going on with this PR. Poor description:

Whenever I try on coordinator:
```
begin;
update test set y = 0;
```

separate session:
```
delete from test;
```

and then try to query citus_lock_waits from a worker, the update session seems to fail.

onderkalaci commented:
> is this expected?

Well, expected wrt to the current code however not from the user's perspective :) With @halilozanakgul we are aware and going to fix this soon.

The problem is that the underlying views rely on `NonCoordinatorNodeList`: https://github.com/citusdata/citus/blob/master/src/backend/distributed/transaction/citus_dist_stat_activity.c#L322

Meaning that, if you are running citus_lock_waits (or citus_dist_stat_activity etc) from the worker, you don't see the transactions originating from the coordinator. However, if you query these views on the coordinator can see this, that's because we have the logic to pull local stat activity: https://github.com/citusdata/citus/blob/master/src/backend/distributed/transaction/citus_dist_stat_activity.c#L331

((or if run the queries from the workers, you can see because `NonCoordinatorNodeList` can reach to all workers))

My guess is that once we introduced citus_dist_stat_activity etc., the coordinator in the metadata concept has not been there. And, as we introduced coordinator in the metadata, we overlooked this part as no one was using it and tests passed with `NonCoordinatorNodeList`.

As implied, the same problem exists for citus_dist_stat_activity, and @halilozanakgul will fix this on his next PR. Once that is fixed, citus_lock_waits would also show these transactions as citus_lock_waits is built on top of citus_dist_stat_activity

marcocitus commented on src/backend/distributed/sql/udfs/citus_lock_waits/latest.sql:
```diff
@@ -21,6 +21,8 @@ citus_dist_stat_activity_with_node_id AS
+ blocking.global_pid as blocking_gpid,
  waiting.pid AS waiting_pid,
  blocking.pid AS blocking_pid,
  waiting.query AS blocked_statement,
```
any thoughts on simplifying this view by e.g. removing hostname and port, or node ID?

marcocitus replied:
It's also not entirely clear to me whether PID refers to the process ID on the worker or the coordinator. Perhaps it doesn't matter either way?

If we could indicate where the block is happening that would be useful.

onderkalaci replied:
> any thoughts on simplifying this view by e.g. removing hostname and port, or node ID?

I'll work on the definition on the view as a follow-up PR.

Currently, we still rely on dist. tx's nodeId. We should rely on the gpid's nodeId, otherwise commands that are not in a tx block may not show the node information properly.

Stay tuned :)

onderkalaci commented on src/backend/distributed/transaction/lock_graph.c:
```diff
@@ -256,13 +362,31 @@ ParseTimestampTzField(PGresult *result, int rowIndex, int colIndex)
 dump_local_wait_edges(PG_FUNCTION_ARGS)
 {
-	WaitGraph *waitGraph = BuildLocalWaitGraph();
+	bool onlyDistributedTx = PG_GETARG_BOOL(0);
```
ops, very good point, thanks
